daemon/c_seccomp/sysinfo.c

```diff
@@ -0,0 +1,406 @@
+err:
+	mem_free0(buf);
+	return -1;
+}
+
+static void
+c_seccomp_print_sysinfo(struct sysinfo *info)
+{
+	TRACE("uptime;    \t %ld", info->uptime); // Seconds since boot
+	TRACE("loads[3];  \t [%ld, %ld, %ld]", info->loads[0], info->loads[1], info->loads[2]);
+	// 1, 5, and 15 minute load averages
+	TRACE("totalram;  \t %ld", info->totalram);  // Total usable main memory size
+	TRACE("freeram;   \t %ld", info->freeram);   // Available memory size
+	TRACE("sharedram; \t %ld", info->sharedram); // Amount of shared memory
+	TRACE("bufferram; \t %ld", info->bufferram); // Memory used by buffers
+	TRACE("totalswap; \t %ld", info->totalswap); // Total swap space size
+	TRACE("freeswap;  \t %ld", info->freeswap);  // Swap space still available
+	TRACE("procs;     \t %hu", info->procs);     // Number of current processes
+	TRACE("totalhigh; \t %ld", info->totalhigh); // Total high memory size
+	TRACE("freehigh;  \t %ld", info->freehigh);  // Available high memory size
+	TRACE("mem_unit;  \t %d", info->mem_unit);   // Memory unit size in bytes
+}
+
+struct sysinfo_fork_data {
+	c_seccomp_t *seccomp;
+	struct sysinfo *info;
+	pid_t target_pid;
+	void *target_datap;
+};
+
+static int
+c_seccomp_do_sysinfo_fork(const void *data)
+{
+	const struct sysinfo_fork_data *sysinfo_params = data;
+	ASSERT(sysinfo_params);
+
+	// unshare mount ns to mount /sys/fs/cgroup if not mounted inside the container
+	if (unshare(CLONE_NEWNS) == -1) {
+		ERROR_ERRNO("Could not unshare mount namespace!");
+		return -1;
+	}
+
+	// mount cgroups if not already mounted by init
+	if (!file_is_mountpoint(CGROUPS_FOLDER)) {
+		if (mount("cgroup2", CGROUPS_FOLDER, "cgroup2",
+			  MS_NOEXEC | MS_NODEV | MS_NOSUID | MS_RELATIME, NULL) == -1 &&
+		    errno != EBUSY) {
+			ERROR_ERRNO("Could not mount cgroups_v2 unified hirachy");
+		}
+		return -1;
+	}
+
+	// initialize struct sysinfo with host values
+	DEBUG("Executing sysinfo in namespaces of container");
+	if (-1 == sysinfo(sysinfo_params->info)) {
+		ERROR_ERRNO("Failed to execute sysinfo");
+		return -1;
+	}
+
+	TRACE("sysinfo struct ns-only!");
+	c_seccomp_print_sysinfo(sysinfo_params->info);
+
+	// overwrite loads from procfs (if lxcfs is enabled, we get the container values)
+	unsigned long load[6] = { 0 };
+
+	char *loadavg = file_read_new("/proc/loadavg", _SC_PAGE_SIZE);
+	if (6 != sscanf(loadavg, "%lu.%lu %lu.%lu %lu.%lu", &load[0], &load[1], &load[2], &load[3],
+			&load[4], &load[5])) {
+		WARN("Could not parse '/proc/loadavg' in namespace of container!");
+	} else {
+		TRACE("parsed /proc/loadavg '%s' %lu-%lu, %lu-%lu, %lu-%lu", loadavg, load[0],
+		      load[1], load[2], load[3], load[4], load[5]);
+
+		sysinfo_params->info->loads[0] =
+			(load[0] << SI_LOAD_SHIFT) + ((load[1] << SI_LOAD_SHIFT) / 100);
+		sysinfo_params->info->loads[1] =
+			(load[2] << SI_LOAD_SHIFT) + ((load[3] << SI_LOAD_SHIFT) / 100);
+		sysinfo_params->info->loads[2] =
+			(load[4] << SI_LOAD_SHIFT) + ((load[5] << SI_LOAD_SHIFT) / 100);
+	}
```

Comment:
The pattern i see here is that, if reading loadavg fails, the host's information is passed through to the container. If this is intended, i'm fine. Or if this "information leak" (not really confidential information) should be avoided, we should insert a dummy value in the error case.

Reply:
we rely on lxcfs here anyway, thus for now I would say. This is sufficient.